## 4.8 Grouping Data & Aggregating Variables

### Contents

### Exercise 4.8 task
### - Importing libraries and files
### - Using groupby() and agg()
### - Analysing the result
### - Creating a Loyalty flag Label using the loc()
### - Deriving Columns with loc()
### - Exporting the dataframe

#### 1. Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

#### Importing libraries and files

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\boss4048\Documents\10-2024 Instacart Basket Analysis'

In [3]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))

In [4]:
ords_prods_merge.head(100)

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0              0             0   2539329        1             1   
1              0             0   2539329        1             1   
2              0             0   2539329        1             1   
3              0             0   2539329        1             1   
4              0             0   2539329        1             1   
..           ...           ...       ...      ...           ...   
95            14            14    738281        2             4   
96            15            15   1673511        2             5   
97            15            15   1673511        2             5   
98            15            15   1673511        2             5   
99            15            15   1673511        2             5   

    orders_day_of_week  order_hour_of_day  total_days_since_last_order  \
0                    2                  8                          NaN   
1                    2                  8                          NaN   
2                    2                  8                          NaN   
3                    2                  8                          NaN   
4                    2                  8                          NaN   
..                 ...                ...                          ...   
95                   2                 10                          8.0   
96                   3                 11                          8.0   
97                   3                 11                          8.0   
98                   3                 11                          8.0   
99                   3                 11                          8.0   

    product_id  add_to_cart_order  ...  Unnamed: 0_y  \
0          196                  1  ...           195   
1        14084                  2  ...         14084   
2        12427                  3  ...         12427   
3        26088                  4  ...         26089   
4        26405                  5  ...         26406   
..         ...                ...  ...           ...   
95       21150                 13  ...         21151   
96       47144                  1  ...         47148   
97        5322                  2  ...          5322   
98       17224                  3  ...         17224   
99       38656                  4  ...         38660   

                                         product_name aisle_id  department_id  \
0                                                Soda       77              7   
1             Organic Unsweetened Vanilla Almond Milk       91             16   
2                                 Original Beef Jerky       23             19   
3                          Aged White Cheddar Popcorn       23             19   
4                    XL Pick-A-Size Paper Towel Rolls       54             17   
..                                                ...      ...            ...   
95                            Fire Grilled Steak Bowl       38              1   
96     Unsweetened Original Almond Breeze Almond Milk       91             16   
97  Gluten Free Dark Chocolate Chunk Chewy with a ...        3             19   
98                   Oats & Honey Gluten Free Granola        3             19   
99  Gluten Free Peanut Butter Dark Chocolate Chewy...        3             19   

    prices  _merge    price_range_loc     busiest_day    Busiest days  \
0      9.0    both  Mid-range product  Regularly busy  Regularly busy   
1     12.5    both  Mid-range product  Regularly busy  Regularly busy   
2      4.4    both  Low-range product  Regularly busy  Regularly busy   
3      4.7    both  Low-range product  Regularly busy  Regularly busy   
4      1.0    both  Low-range product  Regularly busy  Regularly busy   
..     ...     ...                ...             ...             ...   
95     5.9    both  Mid-range product  Regularly busy  Regularly busy   
96    14.0    both  Mid-range product  Regularly busy    Slowest days   
97     

#### 2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [5]:
# Grouping by 'department_id' and 'order_number', and Calculating Mean using agg() function
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

#### 3. Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

In [6]:
# Mean of the whole dataset
ords_prods_merge['order_number'].mean()

17.142304646349487

In [7]:
# Mean of the subset
ords_prods_merge['order_number'][:1000000].mean()

17.313612

Answer: 

After comparing the results, I noticed that out of 21 department_id, 11 department_id from the entire DataFrame have higher mean values than those in the subset DataFrame. On the other hand, the overall mean value of the subset is slightly higher than that of the entire DataFrame.

1. Depeartment_id = 21 has the highest order_number value in both subset and the entire datafrme compared to other department_id.

2. Using sampled data (row 0 - 1,000,000 in this case) for statistical calculations might yield incorrect results.

#### 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

#### Creating a loyalty flag column using the transform()

In [8]:
# Aggregating data with transform() function
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\boss4048\AppData\Local\Temp\ipykernel_24376\186821350.py:2: FutureWarning: The provided callable <function max at 0x000001924819FBA0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [9]:
# This command tells pandas not to assign any options regarding the maximum number of rows to display
pd.options.display.max_rows = None

In [10]:
# Testing the result with transform()
ords_prods_merge[['user_id', 'order_number', 'max_order']].head(100)

user_id  order_number  max_order
0         1             1         10
1         1             1         10
2         1             1         10
3         1             1         10
4         1             1         10
5         1             2         10
6         1             2         10
7         1             2         10
8         1             2         10
9         1             2         10
10        1             2         10
11        1             3         10
12        1             3         10
13        1             3         10
14        1             3         10
15        1             3         10
16        1             4         10
17        1             4         10
18        1             4         10
19        1             4         10
20        1             4         10
21        1             5         10
22        1             5         10
23        1             5         10
24        1             5         10
25        1             5         10
26        1             5         10
27        1             5         10
28        1             5         10
29        1             6         10
30        1             6         10
31        1             6         10
32        1             6         10
33        1             7         10
34        1             7         10
35        1             7         10
36        1             7         10
37        1             7         10
38        1             8         10
39        1             8         10
40        1             8         10
41        1             8         10
42        1             8         10
43        1             8         10
44        1             9         10
45        1             9         10
46        1             9         10
47        1             9         10
48        1             9         10
49        1             9         10
50        1            10         10
51        1            10         10
52        1            10         10
53        1            10         10
54        1            10         10
55        1            10         10
56        1            10         10
57        1            10         10
58        1            10         10
59        2             1         14
60        2             1         14
61        2             1         14
62        2             1         14
63        2             1         14
64        2             1         14
65        2             1         14
66        2             1         14
67        2             1         14
68        2             1         14
69        2             1         14
70        2             1         14
71        2             1         14
72        2             2         14
73        2             2         14
74        2             2         14
75        2             2         14
76        2             2         14
77        2             2         14
78        2             3         14
79        2             3         14
80        2             3         14
81        2             3         14
82        2             3         14
83        2             4         14
84        2             4         14
85        2             4         14
86        2             4         14
87        2             4         14
88        2             4         14
89        2             4         14
90        2             4         14
91        2             4         14
92        2             4         14
93        2             4         14
94        2             4         14
95        2             4         14
96        2             5         14
97        2             5         14
98        2             5         14
99        2             5         14

#### Creating a Loyalty flag Label using the loc()

In [11]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [12]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [13]:
ords_prods_merge.loc[ords_prods_merge['max_order']  <= 10, 'loyalty_flag'] = 'New customer'

In [14]:
# Checking the frequency of the new “loyalty_flag” column using the value_counts() function
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [15]:
# Show top 100 rows by 'user_id', 'loyalty_flag', 'order_number' columns
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(100)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             1
2         1      New customer             1
3         1      New customer             1
4         1      New customer             1
5         1      New customer             2
6         1      New customer             2
7         1      New customer             2
8         1      New customer             2
9         1      New customer             2
10        1      New customer             2
11        1      New customer             3
12        1      New customer             3
13        1      New customer             3
14        1      New customer             3
15        1      New customer             3
16        1      New customer             4
17        1      New customer             4
18        1      New customer             4
19        1      New customer             4
20        1      New customer             4
21        1      New customer             5
22        1      New customer             5
23        1      New customer             5
24        1      New customer             5
25        1      New customer             5
26        1      New customer             5
27        1      New customer             5
28        1      New customer             5
29        1      New customer             6
30        1      New customer             6
31        1      New customer             6
32        1      New customer             6
33        1      New customer             7
34        1      New customer             7
35        1      New customer             7
36        1      New customer             7
37        1      New customer             7
38        1      New customer             8
39        1      New customer             8
40        1      New customer             8
41        1      New customer             8
42        1      New customer             8
43        1      New customer             8
44        1      New customer             9
45        1      New customer             9
46        1      New customer             9
47        1      New customer             9
48        1      New customer             9
49        1      New customer             9
50        1      New customer            10
51        1      New customer            10
52        1      New customer            10
53        1      New customer            10
54        1      New customer            10
55        1      New customer            10
56        1      New customer            10
57        1      New customer            10
58        1      New customer            10
59        2  Regular customer             1
60        2  Regular customer             1
61        2  Regular customer             1
62        2  Regular customer             1
63        2  Regular customer             1
64        2  Regular customer             1
65        2  Regular customer             1
66        2  Regular customer             1
67        2  Regular customer             1
68        2  Regular customer             1
69        2  Regular customer             1
70        2  Regular customer             1
71        2  Regular customer             1
72        2  Regular customer             2
73        2  Regular customer             2
74        2  Regular customer             2
75        2  Regular customer             2
76        2  Regular customer             2
77        2  Regular customer             2
78        2  Regular customer             3
79        2  Regular customer             3
80        2  Regular customer             3
81        2  Regular customer             3
82        2  Regular customer             3
83        2  Regular customer             4
84        2  Regular customer             4
85        2  Regular customer             4
86        2  Regular customer             4
87        2  Regular customer             4
88        2  Regular customer             4
89        2  Regular customer             4


#### 5. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. 
#### Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). 
#### What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.


In [16]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Loyal customers seem to purchase lower-priced products, with an average price of 10.38 dollars, while new customers and regular customers tend to purchase at higher prices, with average prices of 13.29 dollars and 12.49 dollars, respectively.

#### 6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
#### - If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
#### - If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”


In [17]:
# 1. Aggregating Data with transform()
ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\boss4048\AppData\Local\Temp\ipykernel_24376\1007418047.py:2: FutureWarning: The provided callable <function mean at 0x00000192481B45E0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [18]:
# Testing the output of the 'spending_flag' column
ords_prods_merge.head()

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0             0             0   2539329        1             1   
1             0             0   2539329        1             1   
2             0             0   2539329        1             1   
3             0             0   2539329        1             1   
4             0             0   2539329        1             1   

   orders_day_of_week  order_hour_of_day  total_days_since_last_order  \
0                   2                  8                          NaN   
1                   2                  8                          NaN   
2                   2                  8                          NaN   
3                   2                  8                          NaN   
4                   2                  8                          NaN   

   product_id  add_to_cart_order  ...  department_id  prices _merge  \
0         196                  1  ...              7     9.0   both   
1       14084                  2  ...             16    12.5   both   
2       12427                  3  ...             19     4.4   both   
3       26088                  4  ...             19     4.7   both   
4       26405                  5  ...             17     1.0   both   

     price_range_loc     busiest_day    Busiest days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy        Average orders   
1  Mid-range product  Regularly busy  Regularly busy        Average orders   
2  Low-range product  Regularly busy  Regularly busy        Average orders   
3  Low-range product  Regularly busy  Regularly busy        Average orders   
4  Low-range product  Regularly busy  Regularly busy        Average orders   

  max_order  loyalty_flag average_price  
0        10  New customer      6.367797  
1        10  New customer      6.367797  
2        10  New customer      6.367797  
3        10  New customer      6.367797  
4        10  New customer      6.367797  

[5 rows x 24 columns]

In [19]:
# 2. Deriving Columns with loc()
ords_prods_merge.loc[ords_prods_merge['average_price'] < 10 , 'spending_flag'] = 'Low spender'

In [20]:
ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10 , 'spending_flag'] = 'High spender'

In [21]:
# Check the frequency of the new “loyalty_flag” column using the value_counts() function
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [22]:
# Testing the output of spending_flag column
ords_prods_merge[['user_id', 'average_price', 'spending_flag']].head(100)

user_id  average_price spending_flag
0         1       6.367797   Low spender
1         1       6.367797   Low spender
2         1       6.367797   Low spender
3         1       6.367797   Low spender
4         1       6.367797   Low spender
5         1       6.367797   Low spender
6         1       6.367797   Low spender
7         1       6.367797   Low spender
8         1       6.367797   Low spender
9         1       6.367797   Low spender
10        1       6.367797   Low spender
11        1       6.367797   Low spender
12        1       6.367797   Low spender
13        1       6.367797   Low spender
14        1       6.367797   Low spender
15        1       6.367797   Low spender
16        1       6.367797   Low spender
17        1       6.367797   Low spender
18        1       6.367797   Low spender
19        1       6.367797   Low spender
20        1       6.367797   Low spender
21        1       6.367797   Low spender
22        1       6.367797   Low spender
23        1       6.367797   Low spender
24        1       6.367797   Low spender
25        1       6.367797   Low spender
26        1       6.367797   Low spender
27        1       6.367797   Low spender
28        1       6.367797   Low spender
29        1       6.367797   Low spender
30        1       6.367797   Low spender
31        1       6.367797   Low spender
32        1       6.367797   Low spender
33        1       6.367797   Low spender
34        1       6.367797   Low spender
35        1       6.367797   Low spender
36        1       6.367797   Low spender
37        1       6.367797   Low spender
38        1       6.367797   Low spender
39        1       6.367797   Low spender
40        1       6.367797   Low spender
41        1       6.367797   Low spender
42        1       6.367797   Low spender
43        1       6.367797   Low spender
44        1       6.367797   Low spender
45        1       6.367797   Low spender
46        1       6.367797   Low spender
47        1       6.367797   Low spender
48        1       6.367797   Low spender
49        1       6.367797   Low spender
50        1       6.367797   Low spender
51        1       6.367797   Low spender
52        1       6.367797   Low spender
53        1       6.367797   Low spender
54        1       6.367797   Low spender
55        1       6.367797   Low spender
56        1       6.367797   Low spender
57        1       6.367797   Low spender
58        1       6.367797   Low spender
59        2       7.515897   Low spender
60        2       7.515897   Low spender
61        2       7.515897   Low spender
62        2       7.515897   Low spender
63        2       7.515897   Low spender
64        2       7.515897   Low spender
65        2       7.515897   Low spender
66        2       7.515897   Low spender
67        2       7.515897   Low spender
68        2       7.515897   Low spender
69        2       7.515897   Low spender
70        2       7.515897   Low spender
71        2       7.515897   Low spender
72        2       7.515897   Low spender
73        2       7.515897   Low spender
74        2       7.515897   Low spender
75        2       7.515897   Low spender
76        2       7.515897   Low spender
77        2       7.515897   Low spender
78        2       7.515897   Low spender
79        2       7.515897   Low spender
80        2       7.515897   Low spender
81        2       7.515897   Low spender
82        2       7.515897   Low spender
83        2       7.515897   Low spender
84        2       7.515897   Low spender
85        2       7.515897   Low spender
86        2       7.515897   Low spender
87        2       7.515897   Low spender
88        2       7.515897   Low spender
89        2       7.515897   Low spender
90        2       7.515897   Low spender
91        2       7.515897   Low spender
92        2       7.515897   Low spender
93        2       7.515897   Low spender
94        2       7.515897   Low spender
95        2       7.515897   Low spender
96        2       7.515897 

#### 7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
#### - If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
#### - If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
#### - If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [23]:
# 1. Aggregating Data with transform()
ords_prods_merge['median_days'] = ords_prods_merge.groupby(['user_id'])['total_days_since_last_order'].transform(np.median)

C:\Users\boss4048\AppData\Local\Temp\ipykernel_24376\2065529070.py:2: FutureWarning: The provided callable <function median at 0x00000192483A71A0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_days'] = ords_prods_merge.groupby(['user_id'])['total_days_since_last_order'].transform(np.median)


In [24]:
# Testing the new column created
ords_prods_merge.head()

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0             0             0   2539329        1             1   
1             0             0   2539329        1             1   
2             0             0   2539329        1             1   
3             0             0   2539329        1             1   
4             0             0   2539329        1             1   

   orders_day_of_week  order_hour_of_day  total_days_since_last_order  \
0                   2                  8                          NaN   
1                   2                  8                          NaN   
2                   2                  8                          NaN   
3                   2                  8                          NaN   
4                   2                  8                          NaN   

   product_id  add_to_cart_order  ...  _merge    price_range_loc  \
0         196                  1  ...    both  Mid-range product   
1       14084                  2  ...    both  Mid-range product   
2       12427                  3  ...    both  Low-range product   
3       26088                  4  ...    both  Low-range product   
4       26405                  5  ...    both  Low-range product   

      busiest_day    Busiest days  busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy         Average orders         10   
1  Regularly busy  Regularly busy         Average orders         10   
2  Regularly busy  Regularly busy         Average orders         10   
3  Regularly busy  Regularly busy         Average orders         10   
4  Regularly busy  Regularly busy         Average orders         10   

   loyalty_flag average_price spending_flag median_days  
0  New customer      6.367797   Low spender        20.5  
1  New customer      6.367797   Low spender        20.5  
2  New customer      6.367797   Low spender        20.5  
3  New customer      6.367797   Low spender        20.5  
4  New customer      6.367797   Low spender        20.5  

[5 rows x 26 columns]

In [25]:
# 2. Deriving Columns with loc()
ords_prods_merge.loc[ords_prods_merge['median_days'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [26]:
ords_prods_merge.loc[(ords_prods_merge['median_days'] <= 20) & (ords_prods_merge['median_days'] > 10), 'frequency_flag'] = 'Regular customer'

In [27]:
ords_prods_merge.loc[ords_prods_merge['median_days'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [28]:
# Check the frequency of the new “frenquency_flag” column using the value_counts() function
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [29]:
# Checking relevant columns of updated dataframe
ords_prods_merge[['user_id', 'total_days_since_last_order', 'median_days', 'frequency_flag']].head(100)

user_id  total_days_since_last_order  median_days         frequency_flag
0         1                          NaN         20.5  Non-frequent customer
1         1                          NaN         20.5  Non-frequent customer
2         1                          NaN         20.5  Non-frequent customer
3         1                          NaN         20.5  Non-frequent customer
4         1                          NaN         20.5  Non-frequent customer
5         1                         15.0         20.5  Non-frequent customer
6         1                         15.0         20.5  Non-frequent customer
7         1                         15.0         20.5  Non-frequent customer
8         1                         15.0         20.5  Non-frequent customer
9         1                         15.0         20.5  Non-frequent customer
10        1                         15.0         20.5  Non-frequent customer
11        1                         21.0         20.5  Non-frequent customer
12        1                         21.0         20.5  Non-frequent customer
13        1                         21.0         20.5  Non-frequent customer
14        1                         21.0         20.5  Non-frequent customer
15        1                         21.0         20.5  Non-frequent customer
16        1                         29.0         20.5  Non-frequent customer
17        1                         29.0         20.5  Non-frequent customer
18        1                         29.0         20.5  Non-frequent customer
19        1                         29.0         20.5  Non-frequent customer
20        1                         29.0         20.5  Non-frequent customer
21        1                         28.0         20.5  Non-frequent customer
22        1                         28.0         20.5  Non-frequent customer
23        1                         28.0         20.5  Non-frequent customer
24        1                         28.0         20.5  Non-frequent customer
25        1                         28.0         20.5  Non-frequent customer
26        1                         28.0         20.5  Non-frequent customer
27        1                         28.0         20.5  Non-frequent customer
28        1                         28.0         20.5  Non-frequent customer
29        1                         19.0         20.5  Non-frequent customer
30        1                         19.0         20.5  Non-frequent customer
31        1                         19.0         20.5  Non-frequent customer
32        1                         19.0         20.5  Non-frequent customer
33        1                         20.0         20.5  Non-frequent customer
34        1                         20.0         20.5  Non-frequent customer
35        1                         20.0         20.5  Non-frequent customer
36        1                         20.0         20.5  Non-frequent customer
37        1                         20.0         20.5  Non-frequent customer
38        1                         14.0         20.5  Non-frequent customer
39        1                         14.0         20.5  Non-frequent customer
40        1                         14.0         20.5  Non-frequent customer
41        1                         14.0         20.5  Non-frequent customer
42        1                         14.0         20.5  Non-frequent customer
43        1                         14.0         20.5  Non-frequent customer
44        1                          0.0         20.5  Non-frequent customer
45        1                          0.0         20.5  Non-frequent customer
46        1                          0.0         20.5  Non-frequent customer
47        1                          0.0         20.5  Non-frequent customer
48        1                          0.0         20.5  Non-frequent customer
49        1                          0.0         20.5  Non-frequent customer
50        1                         30.0         20.5  Non-frequent customer


#### 8. Ensure your notebook is clean and structured and that your code is well commented.

The notebook looks clean and well structured with comments.

#### 9. Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [30]:
# Saving the file in .pkl format because it can handle well with very large dataset
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ord_prods_grouped_aggregated.pkl'))